### DISCUSSION TITLE
Update related entities in cache

### DISCUSSION THREAD

**Discussion by sarink** [urql-graphql/urql#2694] [Q&A] [answered]
Assume the following schema:
```
type Store {
  id: ID!
  name: String!
  products: [Product!]!
}

type Product {
  id: ID!
  upc: String!
  store: Store!
}

type Query {
  getStores: [Store!]!
}

type Mutation {
  moveProduct(productId: ID!, newStoreId: ID!): Product!
}
```

If a client wants to move a product from store 1 to store 2, they can execute the `moveProduct` mutation, will get back a new product with an updated `product.store`, and it's good to go.

However, the old store (store 1) still has this product in its `products` list. How can we write an `updates` function that fixes this?

**Comment by sarink** [comment]
Currently we have,
```
updateProduct: (result, { storeId: newStoreId }, cache: any) => {
  const productId = result.updateProduct.id;

  const cacheData = [...cache.data.records.base.values()];
  const allStores = cacheData.filter((item) => item.__typename === 'Store') as Store[];
  const newStore = allStores.find((store) => store.id === newStoreId);
  const oldStore = allStores.find((store) => {
    const isNotTheNewStore = store.id !== newStoreId;
    const hasThisProduct = store.products?.find((p) => p.id === productId);
    return isNotTheNewStore && hasThisProduct;
  });

  const UpdatesUpdateProductFragmentDoc = gql`
    fragment UpdatesUpdateProduct on Product {
      id
      store {
        id
        products {
          id
        }
      }
    }
  `;

  // Disconnect the product from its previous store.products[]
  if (oldStore && oldStore.products) {
    cache.writeFragment<UpdatesUpdateProductFragment>(
      UpdatesUpdateProductFragmentDoc,
      {
        id: productId,
        store: {
          id: oldStore.id,
          products: oldStore.products.filter((p) => p.id !== productId),
        },
      }
    );
  }

  // Add the product to its new store.products[]
  cache.writeFragment<UpdatesUpdateProductFragment>(
    UpdatesUpdateProductFragmentDoc,
    {
      id: productId,
      store: {
        id: newStoreId,
        products: [...(newStore?.products ?? []), { id: productId }],
      },
    }
  );
},
```
Which feels like a lot of brittle code and a hack. Is there a better way?

**Reply by kitten** [reply]
I definitely see what you mean by brittle. I'd assume if you were to ship this code you could extract the part that extracts entities from Graphcache's internals just to be sure ...

To be clear, we'd consider these data structures stable and any changes as breaking. That's because this isn't the first time I've seen someone do this.
So, I can at least guarantee you that if this code broke in an update, it'd be a major and our changeling would call the data structure changes out explicitly.

That said, we still don't have a public API for this. We're a little concerned about the performance characteristics of doing this on stores of unknown sizes (You're of course in a better position to estimate this for yourself)
But we still think eventually we'll implement an API to get all keys for a given type.

I'd assume that you'd want to instead check `cache resolve(productKey, 'store')` to get the reverse relationship.

If you aren't sure whether that exists you could try invalidating some data instead.

Alternatively, you could also make an edge type, i.e. a type that both `Store.products` (as a list) and `Product.store` (as a single type) would point to that has a global ID that you can then use to find what you need more easily and invalidate it.

But if you want to avoid added requests, the former suggestion (if possible), or what you have, are the only two options ✌️

**Reply by kitten** [reply]
By the way, if you use the full entry on `const cacheData = [...cache.data.records.base.values()];` then you also get the key from these values. This means that you'll have all the items' store keys (as per `cacheExchange({ keys })` basically.

In turn, checking `__typename` is always safe basically (although `cache.resolve(key, '__typename')` is another approach), but if you then continue using `cache.resolve(key, prop)` for anything else, you'll make this code a little safer.

If we are to implement `cache.keysOfType('TypeName')` then it'd likely also take our layers into account, do some deduplication checks, and return an iterator potentially... which likely all isn't needed if you find the approach based on `cache.data.records.base` sufficient

**Reply by kitten** [reply]
(Also kudos for figuring this out ❤️ 😄 I know being offline-only (rather than offline-first) is in some cases _very_ hard to do with mutations, and requires this exact workaround, but I'm happy you found it!)

**Comment by kitten** [comment, marked as answer]
There are several ways to go about this, and I can't really tell you which one is better by just seeing what you ended up with necessarily 😅 
It basically depends on whether you can guarantee that your lists are always going to be 100% complete. If they are, then it's indeed not a bad idea to update the lists.

In fact, if you're assuming that you have:
- the product cached
- the old and new store cached (at least once the mutation completes)
- and the lists are always complete

Then I'd indeed update the old list, new list, and make sure that the product is updated. (Although, you could also assume that the mutation should do this)

This was partly the basis of our `populateExchange` effort that we never completed. (See #964)
Basically, because you're running the mutation, the mutation could select:
- the product
- the product's store
- the product's store's list of products
If the above is true, then you only need to update the old product's list and you're done. We were basically looking at automating some of these selections with an exchange... but we never got to finishing it.

So that gives you the choice of manually doing the same.

As per the old product, I think your approach is good! There's not really a simple way I see to optimie this, unless you'd want to use `cache.invalidate` instead, in which case it may make sense to have a "connection" (As in Relay-style connection) on the store that you could invalidate separately.

You could however make your code a little shorter by using `cache.link` rather than `writeFragment` 👀  That'd allow you to do a targeted update of just the list: https://formidable.com/open-source/urql/docs/api/graphcache/#link

**Reply by sarink** [reply]
It felt like a hack particularly because of the mechanism used for finding the "old" store - is this the best way though?

**Reply by kitten** [reply]
It _might_ be the best way, but I'm not sure 😅 hence I wrote [this comment](https://github.com/FormidableLabs/urql/discussions/2694#discussioncomment-3704675) to try to put some more suggestions and clarification behind it in relation to your code and listing out all entities

**Comment by sarink** [comment]
The above actually doesn't work, because entities pulled from `cache.data.records.base` do not have any related entities associated with them (that is, my `stores` do not have a `products` array, so it's impossible to find `oldStore`).

However, I was simply able to replace the `allStores` line, and everything else worked as-is. :
```
const allStores = cache.readQuery({
  query: gql`
    query UpdatesUpdateProductGetAllStores {
      getAllStores {
        id
        products {
          id
        }
      }
    }
  `,
})?.getStores ?? [];
```

 I will note that this only works because we have a `getAllStores` query, if we didn't, I'm not sure how you could achieve this cc @kitten ?
